feat(chart): Add `<ChartDataProvider />`

williaster opened the pull request:
🏆 Enhancements

This PR adds a new `<ChartDataProvider />` component to `@superset-ui/charts` package, with storybook demo for a handful of vis plugins (new + old). This is essentially a convenience `React` component around `ChartClient` with an API inspired by the [Apollo `<Query />`](https://www.apollographql.com/docs/react/essentials/queries.html#basic) component:

```jsx
<ChartDataProvider 
  client={SupersetClient} 
  formData={formData} 
  onLoaded={onLoaded} 
  onError={onError}
>
  {({ loading, payload, error }) => (
    {loading && <Loader />}

    {error && <ErrorMessage error={error} />}

    {payload && (
      <SuperChart
        chartType={payload.formData.viz_type}
        chartProps={
          new ChartProps({
            formData: payload.formData,
            height,
            payload: payload.queryData
            width,
          })
        }
      />
    )}
  )}
</ChartDataProvider>
```

![ChartDataProvider](https://user-images.githubusercontent.com/4496521/54467802-b9ad4780-4744-11e9-94b4-85ecb238e622.gif)

The PR makes a few tweaks which were needed to get things working
- [SupersetClient] 🐛 Fixes a reference in the `SupersetClient` public interface where `request` referenced `get`
- [ChartClient]
  - Fixes some types in `ChartClient` that were giving me issues when using them (such as `SliceIdAndOrFormData`)
- [SuperChart]
  - Removes the remaining `.jsx` test file and uses `@ts-ignore` for that test in the `.tsx` file

**Remaining TODO**
- [x] Update `README` to final API
- [x] verify 100% test coverage
- [x] ~separate `mocks/formData` into separate files~ I think given that this is a fixture file now, it's fine in one file

@kristw @conglei @xtinec

williaster commented:
Still have some todos, but I think reviewable enough/the API is stable.

kristw commented on packages/superset-ui-chart/src/components/ChartDataProvider.tsx:
```diff
@@ -0,0 +1,147 @@
+import { ChartFormData } from '../types/ChartFormData';
+import { QueryData } from '../models/ChartProps';
+
+type Datasource = object;
```
Do we want to be more specific here?

williaster replied:
`ChartClient` never really defined a specific `Datasource` type so I was just using that logic. But I agree it'd be useful so I updated it to be more specific in `types/Datasource` (erring on the side of under-defining for now, but using the seemingly most-important fields from an example API response)

kristw replied:
ok

kristw commented on packages/superset-ui-chart/src/components/ChartDataProvider.tsx:
```diff
@@ -0,0 +1,147 @@
+  };
+
+type State = {
+  status: 'unitialized' | 'fetching' | 'error' | 'loaded';
```
typo `unitialized` => `uninitialized`

kristw replied:
Should we defined enum for `status`?

kristw replied:
Another minor: `loading` & `loaded` vs. `fetching` & `loaded`?

williaster replied:
thanks for catching the typos. `loading` + `loaded` seem good instead of the `fetching` inconsistency.

I thought generally we were avoiding enums? (esp here seems overkill vs simple union since it's scoped to the component)

kristw replied:
sure. no strong opinion on `enum`. strings with type checked seems good enough for me.

kristw commented on packages/superset-ui-chart/src/components/ChartDataProvider.tsx:
```diff
@@ -0,0 +1,147 @@
+
+    const provided: ProvidedProps = {};
+
+    if (status === 'unitialized') {
```
typo

kristw commented on packages/superset-ui-chart/README.md:
```diff
@@ -7,15 +7,72 @@ Description
+   configuration in the `@superset-ui` storybook.
 
-`fn(args)`
+3. Register the needeed `@superset-ui` chart + color plugins.
```
typo `needed`

williaster commented:
all right this should be good to go now:

- need to fix `@superset-ui/core` declaration in `@superset-ui/chart` (separate PR)
- moved the declaration of `types/`
- added tests to `ChartClient` for legacy API calls
- set `buildQuery=undefined` by default so TS is clean
- fixed typos (thank you! 🙏 ) + updated readme.
- got up to 💯 % coverage
